### ISSUE TITLE
Standups: Recurrence MVP implementation

### ISSUE THREAD

**Issue by BartoszJarocki** [ParabolInc/parabol#6852]
Earlier discussion reference: https://github.com/ParabolInc/parabol/issues/5916
Design: https://github.com/ParabolInc/parabol/issues/6747 

AC:
- [ ] an option to start meeting automatically is visible on the new meeting screen for standups
- [ ] meeting starts automatically every weekday if enabled
- [ ] previous meeting closes automatically when new one is started

**Comment by BartoszJarocki** [comment]
Possible solution 

## Backend

1. Introduce new table - `MeetingSeries`

When a recurrent meeting is created, a new entry in `MeetingSeries` is created with `NewMeeting` having reference to it (1:n relation, `MeetingSeries` can have N meetings, where `NewMeeting` can be assigned only to 1 `MeetingSeries`)

| id | title? | isRecurring | recurrence pattern |
| --- | --- | --- | --- | 

This requires changing the `NewMeeting` schema to have `meetingSeriesId` reference

q: how can we support recurrence patterns? what format?
- cron?  for example 0 0 * * 1-5 ?
- something like iCal? https://www.kanzaki.com/docs/ical/rrule.html ?
- something simple that would serve us? simple enum with values like weekdays, weekly, biweekly, monthly? 

2. Creating a recurring meeting

When a recurrent meeting is created, we'd also create a `ScheduledJob` instance with `PROCESS_RECURRENCE` type. The job would have given `meetingId` and `runAt` set to current date + processed recurrence pattern (for example, if it's weekdays it'd be next day, except for Sat, Sun). When the job is run, the previous meeting will be closed and another one will be created automatically. 

3. Mutations

- `startRecurring`
- `stopRecurring`

`startRecurring` would take `meetingId` and the `recurrencePattern` (corn, ical or enum, whatever will be finally decided) params. 

### An example

A user creates a recurring Async Standup meeting

1. frontend executed startTeamPrompt with a new parameter, recurrence settings
2. new entry in `MeetingSeries` is created
3. new `NewMeeting` is created with `meetingSeriesId` set to the newly created `MeetingSeries`
4. new `ScheduledJob` is created with `runAt` set to current date + parsed recurrence settings
5. when created above `ScheduledJob` is run, the old meeting is closed, and a new one is created; point 4 is executed again

## Frontend

Nothing really changes for the MVP purposes, if the meeting has a linked meeting series, we'll show the option to start/stop the recurrence which will respectively call start/stop recurrence mutations. On the meeting dashboard, we'll still show the meeting instance, not the meeting series. See the design here: https://github.com/ParabolInc/parabol/issues/6747 

## Concerns
1. What if a job fails? how to recover from that and not disrupt the user experience? 

## Questions
1. Why it wouldn't work? What are the weak points? What am I missing?
2. Is there anything we can do better here?

another possible approach: https://www.notion.so/parabol/Recurrence-draft-702693dab9394ca5ab1e66993cd9d91b 
I do think it sounds good but there are a lot of questions left. For now, we're constrained by time (we have to deliver it by EOM) and I think it's better to deliver a solution that'd work for current needs. 

reading material 
https://github.com/bmoeskau/Extensible/blob/master/recurrence-overview.md 
https://stackoverflow.com/questions/85699/whats-the-best-way-to-model-recurring-events-in-a-calendar-application 

I'd love to hear your thoughts about it @mattkrick @Dschoordsch @jmtaber129 @igorlesnenko, thank you!

**Comment by jmtaber129** [comment]
I think using `ScheduledJob` would be too brittle, since any errors that cause the next job to not be enqueued would break the entire meeting series, and we'd need to introduce additional complexity to recover when this happens (and it _will_ happen).  Instead of using that, I think it makes sense to create a single `processRecurrence` mutation, and run that mutation regularly (e.g. every 5 minutes) via the [chronos microservice](https://github.com/ParabolInc/action-chronos/blob/master/src/jobs/index.js).  This would have the added benefit of allowing overlapping meetings (e.g. a daily standup that's open for 36 hours), and meetings that end before the next one begins.   This mutation would:
* Fetch all active meeting series
* For each meeting series
  * Close any meetings that should be started
  * End any meetings that should be ended

Importantly, this mutation would need to be idempotent to enforce consistency and avoid timing issues (e.g. running the meeting twice in quick succession shouldn't create an extra meeting or close a meeting that was just started).  As such, we'd need a good way to determine whether a meeting should be closed and whether a meeting should be open:
* For closing meetings - check a new `scheduledEndTime` field on the active meetings in the series (for closing meetings) and close if `now` is after that
  * This could also be decoupled from recurrence such that a user could create a single standup that closes automatically after some time.
* For starting meetings - check the most recently started meeting, and determine all desired meeting start times since then (usually just a single point in time, but could be multiple if we need to backfill meetings for any reason).  Start a meeting for each of these start times (assuming the would-be end-time is after `now`).

This mutation might also need a lock to ensure only one instance of the mutation is being run at a given time, since this procedure would be prone to race conditions if multiple instances are run in parallel.

We'd also need to consider how updates to the recurrence expression would impact when meetings start/end, but I think this would be easily solvable with some special-casing in the `processRecurrence` mutation (e.g. track when a meeting series was created+updated and only start meetings that should have started after the last update).

Re: recurrence expression format - we should _not_ re-invent the wheel. Let's pull from existing standards (such as iCal) and use one of those.

(FWIW I only skimmed the Extensible doc, but it does seem like wisdom worth following)

**Comment by mattkrick** [comment]
My .02:
- Looking forward to structured conversations,the thinking today is we're going to have a "Flow" that creates "Activities" (ie Meetings) on some "trigger" (ie cron, webhook, etc). It'd definitely be a premature optimization to build support for this right now, however if we do this right, a "MeetingSeries" could be refactored into a "Flow" in the future. Keeping the name "MeetingSeries" is fine, after all renaming a DB table isn't hard, I just want you to have all the information I have as you make future design decisions
  - Rough design of structured convos: https://www.figma.com/file/C7mTy83hN3qf2552EvgFCm/Structured-Convos?node-id=2%3A277
  - Another concept: https://www.figma.com/file/viS2BD56myBGaLuxwEDnxg/Better-Donut?node-id=47%3A246
- I don't understand how `startRecurring` accepts a `meetingId` since it will be creating meetings (and thus `meetingIds`). I would assume the signature would look like `startRecurring(teamId: ID!, meetingType: MeetingTYpeEnum!, startAt: Date!, runPattern: RunPatternEnum!)`. It would create a new job that would run at `startAt` and when it ran it would create the new meeting. It would create a 2nd job that would run 24 hours after `startAt` which would end the meeting & start a new one. If you have other thoughts let me know! Could be wrong
- @jmtaber129 `processRecurrence` sounds very similar to the existing `runScheduledJobs` cron task (https://github.com/ParabolInc/action-chronos/blob/master/src/jobs/index.js#L62). Every 10 minutes it calls a `runScheduledJobs` which finds every job to be run from the beginning of time up until 10 minutes into the future. It schedules those into the node event loop to be run & once the job is run (ie the meeting is created) the job gets removed (https://github.com/ParabolInc/parabol/blob/6af4493238d4a4ec5948897bc09c830fb26f0739/packages/server/graphql/private/mutations/runScheduledJobs.ts#L46). Would that mutation work for this application as well?
- @BartoszJarocki if James is OK with reusing `runScheduledJobs` perhaps we only delete the job from the table _after_ it completes successfully instead of removing it & then assuming the job completes successfully (see https://github.com/ParabolInc/parabol/blob/6af4493238d4a4ec5948897bc09c830fb26f0739/packages/server/graphql/private/mutations/runScheduledJobs.ts#L48). You could work with @adaniels-parabol to make sure we get an angry alert via slack from datadog if `runScheduledJobs` hasn't run within the last 15 minutes.

**Comment by jmtaber129** [comment]
I think only deleting the job from the table after completing successfully would still leave us open to bugs and other failures.  For example, if the previous meeting is successfully closed and the new meeting is created, but something else fails (e.g. a notification fails to send, etc.), we'd probably close this new meeting and start another new meeting on the new run (something guaranteed not to happen with an idempotent mutation).  I think there are some other downsides, too:
* **Complexity when modifying meeting series recurrence** - if a job is scheduled to close a meeting and start a new one, what happens if the recurrence expression is changed, or if recurrence is cancelled?  We'd need either find the scheduled job and delete it / update it
* **Inflexibility with when meetings start + end** - this approach requires that meetings start+end at the same time.  We could decouple this by scheduling separate jobs for the next meeting start and the current meeting end, but that would also introduce otherwise unnecessary complexity.

The main benefit I see to using a single idempotent `processRecurrence` mutation is that it **keeps the meeting series as the single source of truth** for recurrence state, and as a result **drastically reduces complexity** - if the recurrence settings change on the meeting series, it's guaranteed that those changes will be used going forward; there's no job scheduling, etc. that we need to ensure gets updated, too.  Using scheduled jobs would introduce a lot more moving parts that would be difficult to observe, difficult to debug, and difficult to change (e.g. changing supported recurrence settings would require changes to how jobs get scheduled [possibly including a migration on scheduled jobs] _and_ the process recurrence job itself).

Another way of thinking about this is that scheduled jobs rows represent denormalized recurrence events that need to be kept in-sync with the meeting series, vs. creating a new cron job that calculates these recurrence events directly from the meeting series when we actually process the recurrence.

**Comment by jmtaber129** [comment]
@BartoszJarocki @mattkrick thoughts?  We'll probably want to start executing on this soon

**Comment by BartoszJarocki** [comment]
@jmtaber129 @mattkrick thank you both for such a thoughtful conversation! 

> I don't understand how startRecurring accepts a meetingId since it will be creating meetings (and thus meetingIds). I would assume the signature would look like startRecurring(teamId: ID!, meetingType: MeetingTYpeEnum!, startAt: Date!, runPattern: RunPatternEnum!). It would create a new job that would run at startAt and when it ran it would create the new meeting. It would create a 2nd job that would run 24 hours after startAt which would end the meeting & start a new one. If you have other thoughts let me know! Could be wrong

@mattkrick you're right! The case I had in mind, is that currently we'll show separate meetings in the dashboard and there's an option to start recurrence from the existing meeting instance so I wanted to have `meetingId` to be able to reference the newly created `MeetingSeries` in it. I think it's not that important and we can either remove the option for starting the recurrence from the existing meeting or add an optional `meetingId` param, what do you think? 

@jmtaber129 I think your arguments are very solid and I'd go with your solution. Would you like to compile your thoughts into a separate backend recurrence implementation task? I can do it as well but I'm not sure if I catch all the details you mentioned!

**Comment by jmtaber129** [comment]
@BartoszJarocki yep! I can convert my thoughts into more of a technical spec later today or tomorrow.

**Comment by BartoszJarocki** [comment]
Closing this in favor of the whole epic: https://github.com/ParabolInc/parabol/milestone/141
